clients/ts/signalr/src/LongPollingTransport.ts

```diff
@@ -0,0 +1,174 @@
+import { ILogger, LogLevel } from "./ILogger";
+import { ITransport, TransferFormat } from "./ITransport";
+import { Arg, getDataDetail, sendMessage } from "./Utils";
+
+const SHUTDOWN_TIMEOUT = 5 * 1000;
+
+export class LongPollingTransport implements ITransport {
+    private readonly httpClient: HttpClient;
+    private readonly accessTokenFactory: () => string | Promise<string>;
+    private readonly logger: ILogger;
+    private readonly logMessageContent: boolean;
+
+    private url: string;
+    private pollXhr: XMLHttpRequest;
+    private pollAbort: AbortController;
+    private shutdownTimeout: number;
+    private running: boolean;
+
+    constructor(httpClient: HttpClient, accessTokenFactory: () => string | Promise<string>, logger: ILogger, logMessageContent: boolean) {
+        this.httpClient = httpClient;
+        this.accessTokenFactory = accessTokenFactory || (() => null);
+        this.logger = logger;
+        this.pollAbort = new AbortController();
+        this.logMessageContent = logMessageContent;
+    }
+
+    public connect(url: string, transferFormat: TransferFormat): Promise<void> {
+        Arg.isRequired(url, "url");
+        Arg.isRequired(transferFormat, "transferFormat");
+        Arg.isIn(transferFormat, TransferFormat, "transferFormat");
+
+        this.url = url;
+
+        this.logger.log(LogLevel.Trace, "(LongPolling transport) Connecting");
+
+        if (transferFormat === TransferFormat.Binary && (typeof new XMLHttpRequest().responseType !== "string")) {
+            // This will work if we fix: https://github.com/aspnet/SignalR/issues/742
+            throw new Error("Binary protocols over XmlHttpRequest not implementing advanced features are not supported.");
+        }
+
+        this.poll(this.url, transferFormat);
+        return Promise.resolve();
+    }
+
+    private async poll(url: string, transferFormat: TransferFormat): Promise<void> {
+        this.running = true;
+
+        const pollOptions: HttpRequest = {
+            abortSignal: this.pollAbort.signal,
+            headers: {},
+            timeout: 90000,
+        };
+
+        if (transferFormat === TransferFormat.Binary) {
+            pollOptions.responseType = "arraybuffer";
+        }
+
+        let closeError: Error;
+        try {
+            while (this.running) {
+                // We have to get the access token on each poll, in case it changes
+                const token = await this.accessTokenFactory();
+                if (token) {
+                    // tslint:disable-next-line:no-string-literal
+                    pollOptions.headers["Authorization"] = `Bearer ${token}`;
+                }
+
+                try {
+                    const pollUrl = `${url}&_=${Date.now()}`;
+                    this.logger.log(LogLevel.Trace, `(LongPolling transport) polling: ${pollUrl}`);
+                    const response = await this.httpClient.get(pollUrl, pollOptions);
+
+                    if (!this.running) {
```

Comment:
Why do we need this

Reply:
As I explained elsewhere, this is just diagnostic. I can yank it.